File: src/csharp/4_BehavioralPatterns/1_ChainOfResponsibility/BrokerChain.cs
using System;
using System.Collections.Generic;
using DotNetDesignPatternDemos.Behavioral.ChainOfResponsibility.MethodChain;
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.ChainOfResponsibility.ModifierChain.BrokerChain
{
  // command query separation is being used here

  public class Query
  {
    public string CreatureName;

    public enum Argument
    {
      Attack, Defense
    }

    public Argument WhatToQuery;

    public int Value; // bidirectional

    public Query(string creatureName, Argument whatToQuery, int value)
    {
      CreatureName = creatureName ?? throw new ArgumentNullException(paramName: nameof(creatureName));
      WhatToQuery = whatToQuery;
      Value = value;
    }
  }

  public class Game // mediator pattern
  {
    public event EventHandler<Query> Queries; // effectively a chain

    public void PerformQuery(object sender, Query q)
    {
      Queries?.Invoke(sender, q);
    }
  }

  public class Creature
  {
    private Game game;
    public string Name;
    private int attack, defense;

    public Creature(Game game, string name, int attack, int defense)
    {
      this.game = game ?? throw new ArgumentNullException(paramName: nameof(game));
      this.Name = name ?? throw new ArgumentNullException(paramName: nameof(name));
      this.attack = attack;
      this.defense = defense;
    }

    public int Attack
    {
      get
      {
        var q = new Query(Name, Query.Argument.Attack, attack);
        game.PerformQuery(this, q);
        return q.Value;
      }
    }

    public int Defense
    {
      get
      {
        var q = new Query(Name, Query.Argument.Defense, defense);
        game.PerformQuery(this, q);
        return q.Value;
      }
    }

    public override string ToString() // no game
    {
      return $"{nameof(Name)}: {Name}, {nameof(attack)}: {Attack}, {nameof(defense)}: {Defense}";
      //                                                 ^^^^^^^^ using a property    ^^^^^^^^^
    }
  }

  public abstract class CreatureModifier : IDisposable
  {
    protected Game game;
    protected Creature creature;

    protected CreatureModifier(Game game, Creature creature)
    {
      this.game = game;
      this.creature = creature;
      game.Queries += Handle;
    }

    protected abstract void Handle(object sender, Query q);

    public void Dispose()
    {
      game.Queries -= Handle;
    }
  }

  public class DoubleAttackModifier : CreatureModifier
  {
    public DoubleAttackModifier(Game game, Creature creature) : base(game, creature)
    {
    }

    protected override void Handle(object sender, Query q)
    {
      if (q.CreatureName == creature.Name &&
          q.WhatToQuery == Query.Argument.Attack)
        q.Value *= 2;
    }
  }

  public class IncreaseDefenseModifier : CreatureModifier
  {
    public IncreaseDefenseModifier(Game game, Creature creature) : base(game, creature)
    {
    }

    protected override void Handle(object sender, Query q)
    {
      if (q.CreatureName == creature.Name &&
          q.WhatToQuery == Query.Argument.Defense)
        q.Value += 2;
    }
  }

  public class Demo
  {
    public static void Main()
    {
      var game = new Game();
      var goblin = new Creature(game, "Strong Goblin", 3, 3);
      WriteLine(goblin);

      using (new DoubleAttackModifier(game, goblin))
      {
        WriteLine(goblin);
        using (new IncreaseDefenseModifier(game, goblin))
        {
          WriteLine(goblin);
        }
      }

      WriteLine(goblin);
    }
  }
}

File: src/swift/4_BehavioralPatterns/1_ChainOfResponsibility/BrokerChain.swift
import Foundation

// events required for this! (Observer pattern)
protocol Invocable : class
{
  func invoke(_ data: Any)
}

public protocol Disposable
{
  func dispose()
}

public class Event<T>
{
  public typealias EventHandler = (T) -> ()

  var eventHandlers = [Invocable]()

  public func raise(_ data: T)
  {
    for handler in self.eventHandlers
    {
      handler.invoke(data)
    }
  }

  public func addHandler<U: AnyObject>
    (target: U, handler: @escaping (U) -> EventHandler) -> Disposable
  {
    let subscription = Subscription(target: target, handler: handler, event: self)
    eventHandlers.append(subscription)
    return subscription
  }
}

class Subscription<T: AnyObject, U> : Invocable, Disposable
{
  weak var target: T?
  let handler: (T) -> (U) -> ()
  let event: Event<U>

  init(target: T?, handler: @escaping (T) -> (U) -> (), event: Event<U>)
  {
    self.target = target
    self.handler = handler
    self.event = event
  }

  func invoke(_ data: Any) {
    if let t = target {
      handler(t)(data as! U)
    }
  }

  func dispose()
  {
    event.eventHandlers = event.eventHandlers.filter { $0 as AnyObject? !== self }
  }
}

// cqs = command query separation

class Query
{
  var creatureName: String
  enum Argument
  {
    case attack
    case defense
  }
  var whatToQuery: Argument
  var value: Int

  init(name: String, whatToQuery: Argument, value: Int)
  {
    self.creatureName = name
    self.whatToQuery = whatToQuery
    self.value = value
  }
}

class Game // mediator pattern
{
  let queries = Event<Query>()

  func performQuery(sender: AnyObject?, q: Query)
  {
    queries.raise(q)
  }
}

class Creature : CustomStringConvertible {
  var name: String
  private let _attack: Int
  private let _defense: Int
  private let game: Game

  init(game: Game, name: String, attack: Int, defense: Int) {
    self.game = game
    self.name = name
    _attack = attack
    _defense = defense
  }

  var attack: Int {
    let q = Query(name: name, whatToQuery: Query.Argument.attack, value: _attack)
    game.performQuery(sender: self, q: q)
    return q.value
  }

  var defense: Int {
    let q = Query(name: name, whatToQuery: Query.Argument.defense, value: _defense)
    game.performQuery(sender: self, q: q)
    return q.value
  }

  var description: String {
    return "Name: \(name), A = \(attack), D = \(defense)"
  }
}

class CreatureModifier : Disposable
{
  let game: Game
  let creature: Creature
  var event: Disposable? = nil

  init(_ game: Game, _ creature: Creature)
  {
    self.game = game
    self.creature = creature
    event = self.game.queries.addHandler(target: self, handler: CreatureModifier.handle)
  }

  func handle(q: Query)
  {
    // no abstract classes
    // implemented by descendant
  }

  func dispose()
  {
    event?.dispose()
  }
}

class DoubleAttackModifier : CreatureModifier
{
  override func handle(q: Query)
  {
    if (q.creatureName == creature.name && q.whatToQuery == Query.Argument.attack)
    {
      q.value *= 2
    }
  }
}

class IncreaseDefenseModifier : CreatureModifier
{
  override func handle(q: Query)
  {
    if (q.creatureName == creature.name &&
       q.whatToQuery == Query.Argument.defense)
    {
      q.value += 2
    }
  }
}

func main()
{
  let game = Game()
  let goblin = Creature(game: game, name: "Strong Goblin",
    attack: 3, defense: 3)
  print("Baseline goblin: \(goblin)")
  do
  {
    let _ = DoubleAttackModifier(game, goblin)
    print("Goblin with double attack: \(goblin)")

    do
    {
      let _ = IncreaseDefenseModifier(game, goblin)
      print("Goblin with 2x attack and inc def: \(goblin)")
    }
  }
}

main()